### ISSUE TITLE
Ensemble gets worse results than the best model / Scores don't match paper

### ISSUE THREAD

**Issue by levayz** [iabh1shekbasu/LungCancerDetectionEnsemble#4]
Hi, 

its looks the the ensemble achieves worse results than the best performing model (densenet).
For example the densenet achieves an ROC_AUC of 77.66 on the test set and the ensemble achieves 74.61.
Also the scores don't seem to match the scores presented in your table (especially for the malignant case):

![ensamble_scores](https://github.com/iabh1shekbasu/LungCancerDetectionEnsemble/assets/13402236/069aada1-af9c-4f32-8e4c-ac70930841f7)

Can you help?

Attaching output of the analysis cell:
```
Train CSV's Order densenet169_train.csv, efficientnetb7_train.csv, resnet152_train.csv


Test CSV's Order densenet169_test.csv, efficientnetb7_test.csv, resnet152_test.csv


Training Data Metrics
--------------------------------------------------------------------------------------
densenet169_train.csv efficientnetb7_train.csv resnet152_train.csv
--------------------------------------------------------------------------------------
num models:3
F1 Score [99.40968123 87.1493803  93.44444444]
ROC_AUC Score [99.74187719 89.29631861 98.45055451]
ACC: [99.80721033 96.20204357 97.72508194]


Testing Data Metrics
--------------------------------------------------------------------------------------
densenet169_test.csv efficientnetb7_test.csv resnet152_test.csv
--------------------------------------------------------------------------------------
num models:3
F1 Score [67.76859504 60.4743083  57.26681128]
ROC_AUC Score [77.66010374 74.47840584 71.65052398]
ACC: [90.38224414 87.66954377 87.85450062]


Original Weights
--------------------------------------------------------------------------------------
densenet169_train.csv 1.5196117748236355
efficientnetb7_train.csv 1.4149865215974706
resnet152_train.csv 1.4876690303175615


Ensemble Probabilities with Original Weights [[4.31025991 0.11200743]
 [4.33942841 0.08283901]
 [4.32479042 0.097477  ]
 ...
 [0.96219578 3.46007126]
 [3.42471502 0.99755241]
 [0.29824056 4.12402671]]
--------------------------------------------------------------------------------------


Predictions with Original Weights [0. 0. 0. ... 1. 0. 1.]
--------------------------------------------------------------------------------------


Normalised Weights
--------------------------------------------------------------------------------------
densenet169_train.csv 0.34362729852071566
efficientnetb7_train.csv 0.31996856296812665
resnet152_train.csv 0.3364041385111576


Ensemble Probabilities with Normal Weights [[0.97467195 0.02532806]
 [0.98126777 0.01873225]
 [0.97795771 0.02204231]
 ...
 [0.21757974 0.78242019]
 [0.77442515 0.22557488]
 [0.06744064 0.93255934]]
--------------------------------------------------------------------------------------


Predictions with Normal Weights [0. 0. 0. ... 1. 0. 1.]
--------------------------------------------------------------------------------------


Adjusted Weights
--------------------------------------------------------------------------------------
densenet169_train.csv 0.3585864
efficientnetb7_train.csv 0.344167
resnet152_train.csv 0.2972466


Ensemble Probabilities After Updating Weights [[0.97280076 0.02719924]
 [0.9799696  0.02003042]
 [0.97635128 0.02364874]
 ...
 [0.2166516  0.78334834]
 [0.75754405 0.24245597]
 [0.06527383 0.93472615]]
--------------------------------------------------------------------------------------


Predictions After Updating Weights [0. 0. 0. ... 1. 0. 1.]
--------------------------------------------------------------------------------------




Accuracy (Original) =  89.39580764488286     Total: 1622


Classification Report:
              precision    recall  f1-score   support

      Benign     0.9050    0.9739    0.9382      1340
   Malignant     0.8056    0.5142    0.6277       282

    accuracy                         0.8940      1622
   macro avg     0.8553    0.7440    0.7829      1622
weighted avg     0.8877    0.8940    0.8842      1622

Accuracy 0.8939580764488286
Precision Score 0.8055555555555556
Recall Score 0.5141843971631206
F1 Score 0.6277056277056277
Roc_Auc Score 0.744032497089023
Confusion matrix:
[[1305   35]
 [ 137  145]]

Classwise Accuracy :[0.9738806 0.5141844]


Accuracy (Normal) =  89.39580764488286     Total: 1622


classification metrics for the predictions made with normalized weights
Classification Report:
              precision    recall  f1-score   support

      Benign     0.9050    0.9739    0.9382      1340
   Malignant     0.8056    0.5142    0.6277       282

    accuracy                         0.8940      1622
   macro avg     0.8553    0.7440    0.7829      1622
weighted avg     0.8877    0.8940    0.8842      1622

Accuracy 0.8939580764488286
Precision Score 0.8055555555555556
Recall Score 0.5141843971631206
F1 Score 0.6277056277056277
Roc_Auc Score 0.744032497089023
Confusion matrix:
[[1305   35]
 [ 137  145]]

Classwise Accuracy :[0.9738806 0.5141844]
classification metrics for the predictions made with adjusted weights
Classification Report:
              precision    recall  f1-score   support

      Benign     0.9057    0.9746    0.9389      1340
   Malignant     0.8111    0.5177    0.6320       282

    accuracy                         0.8952      1622
   macro avg     0.8584    0.7462    0.7855      1622
weighted avg     0.8892    0.8952    0.8855      1622

Accuracy 0.8951911220715166
Precision Score 0.8111111111111111
Recall Score 0.5177304964539007
F1 Score 0.6320346320346321
Roc_Auc Score 0.7461786810627712
Confusion matrix:
[[1306   34]
 [ 136  146]]

Classwise Accuracy :[0.97462687 0.5177305 ]
```

**Comment by levayz** [comment]
I think i found a potential bug you have after training the model, during testing (this happens for all the models).
You have the following cell which sets total and correct to 0, then in the next cell you test on the trainloader, but in the cell after that one you test on the testloader but you dont reset the total and correct to 0.
So it looks like the train results and test results are mixed.
Attaching screenshots:

![image](https://github.com/iabh1shekbasu/LungCancerDetectionEnsemble/assets/13402236/a33cadd7-ece4-4fcf-bb99-ff556b90fa21)

![image](https://github.com/iabh1shekbasu/LungCancerDetectionEnsemble/assets/13402236/f7dd4504-642d-4759-8061-428ab5162b0a)

**Comment by levayz** [comment]
After applying the fix above, i ran the metrics and they look much worse :

```
rain CSV's Order densenet169_train.csv, efficientnetb7_train.csv, resnet152_train.csv


Test CSV's Order densenet169_test.csv, efficientnetb7_test.csv, resnet152_test.csv


Training Data Metrics
--------------------------------------------------------------------------------------
densenet169_train.csv efficientnetb7_train.csv resnet152_train.csv
--------------------------------------------------------------------------------------
num models:3
F1 Score [95.13023783 89.6125772  94.38457175]
ROC_AUC Score [98.77139213 95.62980548 98.24044219]
ACC: [98.34200887 96.43339117 98.0913823 ]


Testing Data Metrics
--------------------------------------------------------------------------------------
densenet169_test.csv efficientnetb7_test.csv resnet152_test.csv
--------------------------------------------------------------------------------------
num models:3
F1 Score [31.21387283 62.75303644 47.59615385]
ROC_AUC Score [59.20133376 75.35540383 66.24722134]
ACC: [85.32675709 88.65598027 86.55980271]


Original Weights
--------------------------------------------------------------------------------------
densenet169_train.csv 1.496757943445019
efficientnetb7_train.csv 1.457027254075529
resnet152_train.csv 1.4910890466404925


Ensemble Probabilities with Original Weights [[4.28460048 0.16027361]
 [4.28187269 0.16300151]
 [3.9106431  0.5342312 ]
 ...
 [0.94107031 3.50380404]
 [3.56181901 0.88305507]
 [1.82600668 2.61886772]]
--------------------------------------------------------------------------------------


Predictions with Original Weights [0. 0. 0. ... 1. 0. 1.]
--------------------------------------------------------------------------------------


Normalised Weights
--------------------------------------------------------------------------------------
densenet169_train.csv 0.3367379730509178
efficientnetb7_train.csv 0.32779943234379166
resnet152_train.csv 0.3354625946052906


Ensemble Probabilities with Normal Weights [[0.96394189 0.03605807]
 [0.9633282  0.03667179]
 [0.87980961 0.1201904 ]
 ...
 [0.21172034 0.78827968]
 [0.80133178 0.19866818]
 [0.41081178 0.58918826]]
--------------------------------------------------------------------------------------


Predictions with Normal Weights [0. 0. 0. ... 1. 0. 1.]
--------------------------------------------------------------------------------------


Adjusted Weights
--------------------------------------------------------------------------------------
densenet169_train.csv 0.3585864
efficientnetb7_train.csv 0.344167
resnet152_train.csv 0.2972466


Ensemble Probabilities After Updating Weights [[0.96214219 0.03785778]
 [0.96149753 0.03850246]
 [0.87380899 0.12619102]
 ...
 [0.22205879 0.77794123]
 [0.79141249 0.20858748]
 [0.41742641 0.58257363]]
--------------------------------------------------------------------------------------


Predictions After Updating Weights [0. 0. 0. ... 1. 0. 1.]
--------------------------------------------------------------------------------------




Accuracy (Original) =  86.80641183723797     Total: 1622


Classification Report:
              precision    recall  f1-score   support

      Benign     0.8689    0.9896    0.9253      1340
   Malignant     0.8542    0.2908    0.4339       282

    accuracy                         0.8681      1622
   macro avg     0.8616    0.6402    0.6796      1622
weighted avg     0.8664    0.8681    0.8399      1622

Accuracy 0.8680641183723797
Precision Score 0.8541666666666666
Recall Score 0.2907801418439716
F1 Score 0.43386243386243384
Roc_Auc Score 0.6401661903249708
Confusion matrix:
[[1326   14]
 [ 200   82]]

Classwise Accuracy :[0.98955224 0.29078014]


Accuracy (Normal) =  86.80641183723797     Total: 1622


classification metrics for the predictions made with normalized weights
Classification Report:
              precision    recall  f1-score   support

      Benign     0.8689    0.9896    0.9253      1340
   Malignant     0.8542    0.2908    0.4339       282

    accuracy                         0.8681      1622
   macro avg     0.8616    0.6402    0.6796      1622
weighted avg     0.8664    0.8681    0.8399      1622

Accuracy 0.8680641183723797
Precision Score 0.8541666666666666
Recall Score 0.2907801418439716
F1 Score 0.43386243386243384
Roc_Auc Score 0.6401661903249708
Confusion matrix:
[[1326   14]
 [ 200   82]]

Classwise Accuracy :[0.98955224 0.29078014]
classification metrics for the predictions made with adjusted weights
Classification Report:
              precision    recall  f1-score   support

      Benign     0.8685    0.9903    0.9254      1340
   Malignant     0.8617    0.2872    0.4309       282

    accuracy                         0.8681      1622
   macro avg     0.8651    0.6388    0.6781      1622
weighted avg     0.8673    0.8681    0.8394      1622

Accuracy 0.8680641183723797
Precision Score 0.8617021276595744
Recall Score 0.2872340425531915
F1 Score 0.4308510638297872
Roc_Auc Score 0.638766275007939
Confusion matrix:
[[1327   13]
 [ 201   81]]

Classwise Accuracy :[0.99029851 0.28723404]
```

im attaching the csvs used for this:
[resnet152_test.csv](https://github.com/iabh1shekbasu/LungCancerDetectionEnsemble/files/15318604/resnet152_test.csv)
[resnet152_train.csv](https://github.com/iabh1shekbasu/LungCancerDetectionEnsemble/files/15318605/resnet152_train.csv)
[densenet169_test.csv](https://github.com/iabh1shekbasu/LungCancerDetectionEnsemble/files/15318608/densenet169_test.csv)
[densenet169_train.csv](https://github.com/iabh1shekbasu/LungCancerDetectionEnsemble/files/15318609/densenet169_train.csv)
[efficientnetb7_test.csv](https://github.com/iabh1shekbasu/LungCancerDetectionEnsemble/files/15318610/efficientnetb7_test.csv)
[efficientnetb7_train.csv](https://github.com/iabh1shekbasu/LungCancerDetectionEnsemble/files/15318611/efficientnetb7_train.csv)

**Comment by iabh1shekbasu** [comment]
Dear @levayz,

Firstly, thank you very much for pointing out the potential bug in our testing code. Your observation about the total and correct variables not being reset between testing on the train and test loaders is indeed correct. However, please rest assured that this issue does not affect the results as those two variables play no role in writing the CSV files, which are used for all the analyses.

Regarding the reproducibility issues, the results in our table might be influenced by factors not immediately obvious in the notebook, such as model initialization. Each model run produce different results (was also the case when we were experimenting); therefore, I would ask you to please verify each of the hyperparameters listed in the paper.

Also, please note that this was my first research project and was conducted about three years ago so I apologize for any issues you have faced in reproducing the results. This work was inspired by the following project: https://github.com/Rohit-Kundu/Ensemble-Pneumonia-Detection — do check out this repository

Thank you again for your feedback.

**Comment by bywbilly** [comment]
I met the exact same issue. My results are very similar to yours. And the hyperparameters of the code match that of the paper.
